File: 5-longest-palindromic-substring/longest-palindromic-substring.py
class Solution(object):
    def longestPalindrome(self, s):
        if len(s) < 2:
            return s

        start = 0
        end = 0

        for i in range(len(s)):
            # Odd length palindrome
            l = i
            r = i

            while l >= 0 and r < len(s) and s[l] == s[r]:
                if r - l > end - start:
                    start = l
                    end = r
                l -= 1
                r += 1

            # Even length palindrome
            l = i
            r = i + 1

            while l >= 0 and r < len(s) and s[l] == s[r]:
                if r - l > end - start:
                    start = l
                    end = r
                l -= 1
                r += 1

        return s[start:end + 1]

File: 5-longest-palindromic-substring/longest-palindromic-substring.java
class Solution {
    public String longestPalindrome(String s) {
        int start = 0;
        int maxLen = 1;

        for (int i = 0; i < s.length(); i++) {

            // Odd length
            int l = i;
            int r = i;

            while (l >= 0 && r < s.length() && s.charAt(l) == s.charAt(r)) {
                if (r - l + 1 > maxLen) {
                    start = l;
                    maxLen = r - l + 1;
                }
                l--;
                r++;
            }

            // Even length
            l = i;
            r = i + 1;

            while (l >= 0 && r < s.length() && s.charAt(l) == s.charAt(r)) {
                if (r - l + 1 > maxLen) {
                    start = l;
                    maxLen = r - l + 1;
                }
                l--;
                r++;
            }
        }

        return s.substring(start, start + maxLen);
    }
}